### PR TITLE
AP_HAL: added HAL_BinarySemaphore

### PR DESCRIPTION
**Pull request by tridge**
This adds AP_HAL::BinarySemaphore, which as the name suggests is a binary semaphore. It replaces AP_HAL::EventHandle and AP_HAL::EventSource, fixing the issues found in #25821 
This PR re-works our CAN layers to use the binary semaphore, which simplifies them, removes race conditions, and fixes the CAN layers to be properly event driven in SITL and Linux.
this will also be used by networking in #25741 

In the future I'd like to rename HAL_Semaphore to HAL_Mutex. Calling our mutexes semaphores is misleading

we can get about 800k ops/s on a H743

also saves a few hundred bytes of flash

### CODE REVIEW

**Comment by andyp1per** [comment]
I thought that we remove the use of semaphores so that we got the benefit of priority inheritance in ChibiOS?

**Comment by tridge** [comment]
> I thought that we removed the use of semaphores so that we got the benefit of priority inheritance in ChibiOS?

we need both. We should rename HAL_Semaphore to HAL_Mutex sometime (I didn't do that in this PR, but would like to discuss it on a dev call). We will primarily use mutexes (what we now call HAL_Semaphore), but in some places we need true semaphores. It allow for wait/signal semantics which you can't do with a mutex.
The first user is networking, but I plan on using them to replace events in CAN, and in some other places.
Basically any place where a thread needs to be woken up when there is work to be done we should use a counting semaphore. When we want to protect a data structure we should use a mutex. Often we will use both.

File: libraries/AP_HAL_ChibiOS/Semaphores.h [hunk 1]
```diff
@@ -37,3 +37,17 @@ class ChibiOS::Semaphore : public AP_HAL::Semaphore {
...
+    CountingSemaphore(uint8_t initial_count=0);
+
+    bool wait(uint32_t timeout_us) override;
+    bool wait_blocking(void) override;
+    void signal(void) override;
+    void signal_ISR(void) override;
+    uint8_t get_count(void) override;
+
+protected:
+    uint32_t _lock[5];
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/Semaphores.h:52]: why not include "ch.h" and use semaphore_t and mutex_t. I double checked it seems to be compiling fine with that. Maybe it was done to go around a circular include, it doesn't seem to be the case anymore.

File: libraries/AP_HAL_ChibiOS/CANFDIface.cpp [hunk 2]
```diff
@@ -996,13 +992,10 @@ uint32_t CANIface::getErrorCount() const
 }
 
 #if CH_CFG_USE_EVENTS == TRUE
-ChibiOS::EventSource CANIface::evt_src_;
-bool CANIface::set_event_handle(AP_HAL::EventHandle* handle)
+bool CANIface::set_event_handle(AP_HAL::CountingSemaphore *handle)
 {
-    CriticalSectionLocker lock;
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/CANFDIface.cpp:1002]: check if handle is null, maybe assert and return false as well.

File: libraries/AP_HAL_ChibiOS/CANFDIface.cpp [hunk 3]
```diff
@@ -996,13 +992,10 @@ uint32_t CANIface::getErrorCount() const
 }
 
 #if CH_CFG_USE_EVENTS == TRUE
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/CANFDIface.cpp:994]: Depends on Counting semaphore enabled now. This condition needs to be removed. There might be other places where this is done.

File: libraries/AP_HAL_SITL/Semaphores.cpp [hunk 4]
```diff
@@ -76,4 +76,151 @@ bool Semaphore::take_nonblocking()
...
+}
+
+bool CountingSemaphore::wait_blocking(void)
+{
+    return sem_wait(&_sem) == 0;
+}
+
+void CountingSemaphore::signal(void)
+{
+    sem_post(&_sem);
```
Comment (bugobliterator) [inline libraries/AP_HAL_SITL/Semaphores.cpp:131]: same point as discussed for AP_HAL_ChibiOS::CountinSemaphore::signal. We should probably be limiting the value to 1. So we don't create multiple runs due to multiple signals between waits.

File: libraries/SITL/SIM_Aircraft.cpp [hunk 5]
```diff
@@ -898,7 +898,7 @@ void Aircraft::smooth_sensors(void)
  */
 float Aircraft::filtered_idx(float v, uint8_t idx)
 {
-    if (sitl->servo_speed <= 0) {
+    if (sitl == nullptr || sitl->servo_speed <= 0) {
```
Comment (bugobliterator) [inline libraries/SITL/SIM_Aircraft.cpp:901]: my guess is these changes are so that examples build in SITL. Should we add examples to run under CI, to ensure examples stay working under sitl.

**Review by bugobliterator — CHANGES_REQUESTED** [review]
I have done a first pass at the review, I will do some tests on hardware as well.

File: libraries/AP_HAL_ChibiOS/Semaphores.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.h:52]: for the same reason we use this mechanism for the mutex - it makes building faster. Including ch.h costs quite a lot in a header that ends up being included everywhere. We originally used ch.h in Semaphores.h then moved away from it to speed up builds

File: libraries/AP_HAL_ChibiOS/CANFDIface.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/CANFDIface.cpp:1002]: why is it invalid to set a null handle?

File: libraries/SITL/SIM_Aircraft.cpp (hunk 5 above)
Comment (tridge) [reply libraries/SITL/SIM_Aircraft.cpp:901]: we check they build in CI now, what we don't do is run them. Running them is tricky as they just produce terminal output and run forever, so pass condition isn't obvious

**Comment by andyp1per** [comment]
Looking good. I guess I missed the discussion as to why we don't just wrap the ChibiOS eventing API if what we actually want is an eventing API? The ChibiOS eventing API has far less restrictions and drawbacks.

File: libraries/AP_HAL_ChibiOS/CANFDIface.cpp (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/CANFDIface.cpp:1002]: Yeah, we already check before using everywhere so doesn't matter.

File: libraries/AP_HAL_ChibiOS/Semaphores.cpp [hunk 6]
```diff
@@ -78,4 +80,61 @@ void Semaphore::assert_owner(void)
...
+bool CountingSemaphore::wait_blocking(void)
+{
+    semaphore_t *sem = (semaphore_t *)_lock;
+    return chSemWait(sem) == MSG_OK;
+}
+
+void CountingSemaphore::signal(void)
+{
+    semaphore_t *sem = (semaphore_t *)_lock;
+    chSemSignal(sem);
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.cpp]: I had a look at what networking needs and found that some of the ports use a binary semaphore (eg. posix port) and some use a counting semaphore with no count limit (eg. win32). So I think networking can use a binary semaphore for ArduPilot.
that means we don't actually have a need for a full counting semaphore yet and I've re-implemented it as BinarySemaphore

File: libraries/AP_HAL_ChibiOS/Semaphores.cpp [hunk 7]
```diff
@@ -78,4 +80,61 @@ void Semaphore::assert_owner(void)
...
+}
+
+void CountingSemaphore::signal_ISR(void)
+{
+    semaphore_t *sem = (semaphore_t *)_lock;
+    chSysLockFromISR();
+    chSemSignalI(sem);
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.cpp]: changed to binary semaphore

File: libraries/AP_HAL_ChibiOS/CANFDIface.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/CANFDIface.cpp]: counting semaphores are always enabled, condition removed

File: libraries/AP_HAL_SITL/Semaphores.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_HAL_SITL/Semaphores.cpp]: changed to binary

File: libraries/SITL/SIM_Aircraft.cpp (hunk 5 above)
Comment (tridge) [reply libraries/SITL/SIM_Aircraft.cpp]: i've done it in another way that is less intrusive

**Comment by tridge** [comment]
> I guess I missed the discussion as to why we don't just wrap the ChibiOS eventing API if what we actually want is an eventing API? The ChibiOS eventing API has far less restrictions and drawbacks.

because the ChibiOS eventing API doesn't work (and can't be properly implemented) in Linux, SITL and ESP32.
Plus, we currently do have a wrapper around it in master, and its badly broken. This PR removes it. 
See this PR for explanation:
https://github.com/ArduPilot/ardupilot/pull/25821

File: libraries/AP_HAL_ChibiOS/Semaphores.cpp [hunk 8]
```diff
@@ -78,4 +80,57 @@ void Semaphore::assert_owner(void)
...
+    return chBSemWait(sem) == MSG_OK;
+}
+
+void BinarySemaphore::signal(void)
+{
+    auto *sem = (binary_semaphore_t *)_lock;
+    chBSemSignal(sem);
+}
+
+void BinarySemaphore::signal_ISR(void)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/Semaphores.cpp:126]: I think you need the "I" version of this as well as this version in order to be useful. Usually you know if you are in an ISR and typically will already have a lock.

File: libraries/AP_HAL_ChibiOS/Semaphores.h [hunk 9]
```diff
@@ -37,3 +37,21 @@ class ChibiOS::Semaphore : public AP_HAL::Semaphore {
     // we declare the lock as a uint32_t array, and cast inside the cpp file
     uint32_t _lock[5];
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/Semaphores.h:38]: I think it would be clearer to have something like "#define SIZEOF_BINARY_SEMAPHORE 5", also the uint32_t is presumably for alignment, wouldn't void* be more correct?

File: libraries/AP_HAL_ChibiOS/Semaphores.cpp (hunk 8 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.cpp:126]: none of the use cases so far have the lock. It is typically called as the last thing before you return from the interrupt.
The reason I'm reluctant to have an I version is that is a ChibiOS specific concept, and this is a HAL level call, uniform on SITL, Linux, ESP32 and ChibiOS. If we run across a case where we really need it (which would have to be in a ChibiOS specific driver) then we can address it, or the caller could unlock before making this call

File: libraries/AP_HAL_ChibiOS/Semaphores.h (hunk 9 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.h:38]: we don't actually know the size, as it can vary with compile options (debug, assert etc). We use a static assert in the cpp file to ensure it is big enough

File: libraries/AP_HAL_ChibiOS/Semaphores.cpp (hunk 8 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.cpp:126]: if/when we do have a use case. I think we'd do this:
```
void BinarySemaphore::signal_ISR(void)
{
    auto *sem = (binary_semaphore_t *)_lock;
    const bool was_locked = !port_irq_enabled(port_get_irq_status());
    if (!was_locked) {
      chSysLockFromISR();
    }
    chBSemSignalI(sem);
    if (!was_locked) {
      chSysUnlockFromISR();
   }
}
```
that is what we do in hrt.c for hrt_micros64()

File: libraries/AP_HAL_ChibiOS/Semaphores.h (hunk 9 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Semaphores.h:38]: look further up in Semaphores.h for the exact same pattern for a mutex

**Comment by tridge** [comment]
rebased

File: libraries/AP_HAL_SITL/Semaphores.cpp [hunk 10]
```diff
@@ -76,4 +76,77 @@ bool Semaphore::take_nonblocking()
...
 
+
+/*
+  binary semaphore using pthread condition variables
+ */
+
+BinarySemaphore::BinarySemaphore(bool initial_state) :
+    AP_HAL::BinarySemaphore(initial_state)
+{
+    pthread_cond_init(&cond, NULL);
```
Comment (bugobliterator) [inline libraries/AP_HAL_SITL/Semaphores.cpp:87]: same point as raised in SITL implementation. The reasoning behind the choice of pthread_cond_* api over sem_* api to implement this.

**Review by bugobliterator — CHANGES_REQUESTED** [review]
Minor points, otherwise looks good.

File: libraries/AP_HAL_SITL/Semaphores.cpp (hunk 10 above)
Comment (tridge) [reply libraries/AP_HAL_SITL/Semaphores.cpp:87]: - sem_ doesn't work on MacOSX
- you can't avoid a race in signal that leads to count going over 1 without adding another mutex

File: libraries/AP_HAL_Linux/Semaphores.cpp [hunk 11]
```diff
@@ -44,3 +44,56 @@ bool Semaphore::take_nonblocking()
...
 }
 
+/*
+  binary semaphore using pthread condition variables
+ */
+
+BinarySemaphore::BinarySemaphore(bool initial_state) :
+    AP_HAL::BinarySemaphore(initial_state)
+{
+    pthread_cond_init(&cond, NULL);
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_Linux/Semaphores.cpp:54]: you can't do binary on top of counting without a race without an extra mutex

File: libraries/AP_HAL/examples/BinarySem/BinarySem.cpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    uint32_t ops, timeouts;
+    uint32_t last_print_us;
+    HAL_Semaphore mtx;
+};
+
+void CountSemTest::setup(void)
+{
+    hal.scheduler->thread_create(
+        FUNCTOR_BIND_MEMBER(&CountSemTest::thread1, void), "thd1", 2048, AP_HAL::Scheduler::PRIORITY_IO, 0);
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL/examples/BinarySem/BinarySem.cpp]: fixed

**Comment by tridge** [comment]
@bugobliterator another reason for not using sem_t calls is the behaviour varies with platform. On some the get count call returns -num_waiting, on some it returns 0 when not waiting

**Review by bugobliterator — APPROVED** [review]
